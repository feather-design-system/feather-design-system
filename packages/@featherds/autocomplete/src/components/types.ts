import { Ref, PropType } from "vue";
import HighlightProps from "./Highlight/HighlightProps";
import { InputWrapperProps, InputSubTextProps } from "@featherds/input-helper";
interface IAutocompleteItem extends IAutocompleteItemType {
  _new?: string;
  _text: string;
  _pre?: IAutocompleteChipIcon;
}
interface IAutocompleteItemType {
  [k: string]: unknown;
}

interface IAutocompleteGridColumn {
  prop: keyof IAutocompleteItemType;
  title: string;
  align?: "right" | "left" | "center";
}

interface IAutocompleteResultRender {
  reset: () => void;
  handleKeyPress: (
    e: KeyboardEvent,
    internalResults: IAutocompleteItemType[]
  ) => boolean;
  active: {
    row: number;
    col?: number;
  };
  first: () => void;
}

export enum AutocompleteTypes {
  multi = "multi",
  single = "single",
}
interface IStrategyOptions {
  selectionLimit: Ref<number>;
  selectionLimitReached: Ref<boolean>;
  modelValue: Ref<IAutocompleteItemType | IAutocompleteItemType[]>;
  textProp: Ref<string>;
  input: Ref<HTMLInputElement>;
  query: Ref<string>;
  emitSearch: () => void;
  forceCloseResults: Ref<boolean>;
  allowNew: Ref<boolean>;
  internalResults: Ref<IAutocompleteItemType[]>;
}

interface IAutocompleteChipIcon {
  icon: unknown;
  title: string;
}

export type {
  IStrategyOptions,
  IAutocompleteGridColumn,
  IAutocompleteItemType,
  IAutocompleteItem,
  IAutocompleteChipIcon,
  IAutocompleteResultRender,
};

export const LABELS = {
  noResults: "No results found",
  minChar: "Enter ${min} characters to search",
  clear: "Clear value",
  selectionLimit: "Selection limit reached",
  new: "New",
  remove: "Remove",
};
export const props = {
  id: {
    type: String,
  },
  modelValue: {
    type: [Array, Object] as PropType<
      IAutocompleteItemType[] | IAutocompleteItemType
    >,
  },
  results: {
    type: Array as PropType<IAutocompleteItemType[]>,
    default: () => [],
  },
  textProp: {
    type: String as unknown as PropType<keyof IAutocompleteItemType>,
    default: "_text",
  },
  loading: {
    type: Boolean,
    default: false,
  },
  minChar: {
    type: Number,
    default: 0,
  },
  allowNew: {
    type: Boolean,
    default: false,
  },
  selectionLimit: {
    type: Number,
  },
  newMatcher: {
    type: Function,
    default: (
      item: IAutocompleteItemType,
      query: string,
      comp: { textProp: string }
    ) => {
      return (
        (item[comp.textProp] as string).toString().toLowerCase() ===
        query.toLowerCase()
      );
    },
  },
  type: {
    type: String as PropType<keyof typeof AutocompleteTypes>,
    required: true,
    validator: (v: keyof typeof AutocompleteTypes) => {
      // The value must match either
      return !!AutocompleteTypes[v];
    },
  },
  labels: {
    type: Object as PropType<Partial<typeof LABELS>>,
    default() {
      return LABELS;
    },
  },
  gridConfig: {
    type: Array as PropType<IAutocompleteGridColumn[]>,
  },
  schema: {
    type: Object,
    required: false,
  },
  ...HighlightProps,
  ...InputSubTextProps,
  ...InputWrapperProps,
} as const;

export const emits = {
  "update:modelValue": (
    value: IAutocompleteItemType | IAutocompleteItemType[] | undefined
  ) => true,
  new: (value: string) => true,
  search: (value: string) => true,
};
