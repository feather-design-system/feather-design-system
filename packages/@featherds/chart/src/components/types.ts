export type ShirtSize = "xs" | "sm" | "md" | "lg" | "xl";

export type FeatherChartType = "bar";

export type FeatherBarChartData = {
  labels: Array<string>;
  datasets: Array<{
    label: string;
    data: Array<string | number>;
    backgroundColor?: string;
  }>;
};

export type FeatherBarChartOptions = {
  responsive?: boolean;
  maintainAspectRation?: boolean;
};

export type FeatherD3Dimensions = {
  height: number;
  width: number;
  margin: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  }
};

export type FeatherD3Options = {
  units: string;
  class?: "color1" | "color2";
};


// Data can be any array of objects returned from an API or local JSON
export type FeatherD3Data = Array<unknown>;

export interface ICoordinateAxes {
  x: string,
  y: string
}
