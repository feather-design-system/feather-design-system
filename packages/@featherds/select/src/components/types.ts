export interface ISelectItem extends ISelectItemType {
  _text: string;
}

export interface ISelectItemType {
  [k: string]: unknown;
}
