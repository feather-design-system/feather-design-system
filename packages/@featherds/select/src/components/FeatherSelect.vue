<template>
  <div v-bind="inherittedAttrs" class="feather-select-container">
    <FeatherMenu
      no-expand
      :open="showMenu"
      @outside-click="handleOutsideClick"
      @trigger-click="handleTriggerClick"
      @close="closeMenu"
      class="feather-select-menu-container"
      data-ref-id="feather-select-menu-container"
    >
      <template v-slot:trigger>
        <InputWrapper
          :for="inputId"
          :raised="raised"
          :focused="hasFocus"
          :show-clear="showClear"
          @clear="handleClear"
          menu-trigger
          class="feather-select-wrapper"
          :class="{ focused: hasFocus }"
        >
          <template v-slot:pre>
            <slot name="pre" />
          </template>
          <input
            v-bind="inputAttrs"
            class="feather-select-input"
            data-ref-id="feather-select-input"
            v-on="inputListeners"
            ref="input"
          />
          <template v-slot:post>
            <FeatherIcon
              class="feather-select-icon"
              :class="{ rotate: showMenu }"
              :icon="icon"
            ></FeatherIcon>
          </template>
        </InputWrapper>
      </template>
      <List
        ref="list"
        data-ref-id="feather-select-list"
        :label="label"
        :options="options"
        :text-prop="textProp"
        :active-index="activeIndex"
        @select="handleSelect"
        @keydown="handleKeyDown"
      >
      </List>
    </FeatherMenu>
    <InputSubText :id="subTextId"></InputSubText>
  </div>
</template>
<script lang="ts">
import {
  InputWrapper,
  InputSubText,
  InputWrapperProps,
  InputSubTextProps,
  useInputWrapper,
  useInputSubText,
  useInputInheritAttrs,
} from "@featherds/input-helper";
import { FeatherIcon } from "@featherds/icon";
import KeyboardArrowDown from "@featherds/icon/navigation/ExpandMore";
import { FeatherMenu } from "@featherds/menu";
import { getSafeId } from "@featherds/utils/id";
import { KEYCODES } from "@featherds/utils/keys";
import List from "./List.vue";
import { useValidation } from "@featherds/input-helper";
import { _setTimeout } from "@featherds/utils/setTimeout";
import { computed, defineComponent, PropType, Ref, toRef } from "vue";
import { ISelectItemType } from "./types";
export const props = {
  ...InputWrapperProps,
  ...InputSubTextProps,
  modelValue: {
    type: Object as PropType<ISelectItemType>,
    required: false,
  },
  textProp: {
    type: String,
    default: "_text",
  },
  options: {
    type: Array as PropType<ISelectItemType[]>,
    default: () => {
      return [] as ISelectItemType[];
    },
  },
  schema: {
    type: Object,
    required: false,
  },
} as const;
export const emits = {
  "update:modelValue": (v: ISelectItemType | undefined) => true,
};
export default defineComponent({
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  emits,
  props,
  setup(props, context) {
    useInputSubText(props);
    useInputWrapper(props);
    const inputId = computed(() => {
      return getSafeId("feather-select-input");
    });

    const { validate } = useValidation(
      inputId,
      toRef(props, "modelValue"),
      props.label,
      props.schema as Record<string, unknown>,
      toRef(props, "error") as Ref<string>
    );

    return {
      inputId,
      validate,
      ...useInputInheritAttrs(context.attrs as Record<string, unknown>),
    };
  },
  data() {
    return {
      hasFocus: false,
      showMenu: false,
      charsSoFar: "",
      internalValue: this.modelValue,
      delayTimeout: 0,
    };
  },
  computed: {
    showClear() {
      return !!this.modelValue;
    },
    subTextId() {
      return getSafeId("feather-select-description");
    },
    inputAttrs() {
      return {
        id: this.inputId,
        "aria-haspopup": "listbox" as
          | "grid"
          | "dialog"
          | "menu"
          | "listbox"
          | "tree"
          | undefined
          | boolean
          | "true"
          | "false",
        "aria-invalid":
          (this.$attrs["aria-invalid"] as boolean) || !!this.error,
        value: this.valueText,
        readonly: true,
        disabled: this.disabled,
        "aria-disabled": this.disabled,
        "aria-describedby": ((this.$attrs["aria-describedby"] as string) || "")
          .split(" ")
          .concat([this.subTextId])
          .filter(Boolean)
          .join(" "),
      };
    },
    inputListeners() {
      return {
        focus: this.handleInputFocus,
        blur: this.handleInputBlur,
        keydown: this.handleKeyDown,
      };
    },
    raised() {
      return !!this.internalValue || this.hasFocus;
    },
    valueText() {
      if (this.internalValue && this.internalValue[this.textProp]) {
        return this.internalValue[this.textProp] as string;
      }
      return "";
    },
    activeIndex() {
      if (this.internalValue && this.internalValue[this.textProp]) {
        const iv = this.internalValue as ISelectItemType;
        const found = this.options.filter(
          (x) => x[this.textProp] === iv[this.textProp]
        );
        if (found && found.length) {
          return this.options.indexOf(found[0]);
        }
      }
      return -1;
    },
    icon: () => KeyboardArrowDown,
  },
  watch: {
    showMenu(v) {
      if (v) {
        if (!this.internalValue) {
          this.select(this.options[0]);
        }
        this.$nextTick(() => {
          (this.$refs.input as HTMLInputElement).focus();
        });
      } else {
        this.emitSelection();
      }
    },
    modelValue(v) {
      this.internalValue = v;
    },
  },
  methods: {
    closeMenu() {
      this.showMenu = false;
      this.$nextTick(() => {
        (this.$refs.input as HTMLInputElement).focus();
      });
    },
    handleClear() {
      this.handleSelect(undefined);
      this.emitSelection();
    },
    handleInputFocus() {
      if (this.disabled) {
        return;
      }
      if (!this.hasFocus) {
        this.hasFocus = true;
      }
    },
    handleInputBlur() {
      if (this.hasFocus && !this.showMenu) {
        this.hasFocus = false;
        this.validate();
      }
    },
    handleTriggerClick() {
      if (this.disabled) {
        return;
      }
      this.showMenu = true;
      this.hasFocus = true;
    },
    handleOutsideClick() {
      this.showMenu = false;
      this.hasFocus = false;
    },
    handleSelect(option: ISelectItemType | undefined) {
      this.select(option);
      this.showMenu = false;
      (this.$refs.input as HTMLInputElement).focus();
    },
    select(option: ISelectItemType | undefined) {
      this.internalValue = option;
    },
    emitSelection() {
      this.$emit("update:modelValue", this.internalValue);
    },
    handleKeyDown(e: KeyboardEvent) {
      //enter
      if (e.keyCode === KEYCODES.ENTER) {
        e.preventDefault();
        this.showMenu = !this.showMenu;
        //if hidden focus button
        if (!this.showMenu) {
          this.$nextTick(() => {
            (this.$refs.input as HTMLInputElement).focus();
          });
        }
      }
      //esc
      else if (e.keyCode === KEYCODES.ESCAPE) {
        this.closeMenu();
        e.stopPropagation();
      } else if (e.keyCode === KEYCODES.DOWN) {
        //down
        e.preventDefault();

        if (this.activeIndex + 1 < this.options.length) {
          this.select(this.options[this.activeIndex + 1]);
        }
        this.showMenu = true;
      } else if (e.keyCode === KEYCODES.UP) {
        //up
        e.preventDefault();

        if (this.activeIndex - 1 >= 0) {
          this.select(this.options[this.activeIndex - 1]);
        }
        this.showMenu = true;
      } else if (e.keyCode === KEYCODES.HOME) {
        //home
        e.preventDefault();
        this.select(this.options[0]);
        this.showMenu = true;
      } else if (e.keyCode === KEYCODES.END) {
        //home
        e.preventDefault();
        this.select(this.options[this.options.length - 1]);
        this.showMenu = true;
      } else if (this.showMenu) {
        const char = String.fromCharCode(e.keyCode);
        this.charsSoFar += char;
        this.searchAfterDelay();
      }
    },
    searchAfterDelay() {
      clearTimeout(this.delayTimeout);
      this.delayTimeout = _setTimeout(() => {
        const found = this.options.filter(
          (x) =>
            x[this.textProp] &&
            (x[this.textProp] as string)
              .toLowerCase()
              .indexOf(this.charsSoFar.toLowerCase()) === 0
        );
        if (found && found.length) {
          this.select(found[0]);
        }
        this.charsSoFar = "";
      }, 250);
    },
  },
  components: {
    InputWrapper,
    InputSubText,
    FeatherMenu,
    List,
    FeatherIcon,
  },
});
</script>

<style lang="scss" scoped>
@import "@featherds/styles/mixins/typography";
@import "@featherds/styles/themes/variables";

.feather-select-input {
  border: none;
  padding: 0;
  padding-left: 0.25rem;
  margin: 0;
  display: inline-block;
  vertical-align: middle;
  height: 100%;
  background-color: transparent;
  @include body-small();
  line-height: 1.5em; //IE has an issue with the mixin lineheight when typing
  color: var($secondary-text-on-surface);
  flex: 1;
  pointer-events: none;
  cursor: pointer;
  min-width: 0;
  &:focus,
  &:active {
    border-color: transparent;
    outline: 0;
  }
  &:hover {
    cursor: pointer;
  }
}
.feather-select-icon {
  flex: none;
  font-size: 1.25rem;
  height: 1.25rem;
  width: 1.25rem;
  align-self: center;
  color: var($secondary-text-on-surface);
  transition: transform 280ms ease-in-out;
  transform-origin: center center;
  &.rotate {
    transform: rotate(180deg);
  }
}
.feather-select-menu-container {
  width: 100%;
  position: relative;
  :deep(.feather-menu-dropdown) {
    min-width: 100%;
  }
}
.feather-select-wrapper {
  &.focused .feather-select-input {
    color: var($primary-text-on-surface);
  }
  &.disabled {
    .feather-select-input {
      color: var($disabled-text-on-surface);
    }
    .feather-select-icon {
      color: var($disabled-text-on-surface);
    }
  }
  &:hover {
    cursor: pointer;
    :deep(.feather-input-wrapper) {
      cursor: pointer !important;
    }
  }
  :deep(.feather-input-wrapper) {
    align-items: center;
  }
}
</style>
