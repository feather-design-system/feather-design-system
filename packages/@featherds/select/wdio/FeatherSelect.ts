const OPTION = "[data-ref-id='feather-select-list'] li";
export class FeatherSelect {
  private selectSelector: string;
  get select() {
    return $(this.selectSelector);
  }
  constructor(_select: string) {
    this.selectSelector = _select;
  }

  async selectByText(text: string) {
    const runInBrowser = function (argument: WebdriverIO.Element) {
      argument.click();
    };
    const select = await this.select;
    await browser.execute(runInBrowser, select);

    await $(OPTION).waitForDisplayed({ timeout: 60000 });

    const items = await $$(OPTION);
    const textArray = await Promise.all(items.map((item) => item.getText()));
    const itemIndex = textArray.indexOf(text);
    if (itemIndex > -1) {
      const result = await items[itemIndex].getText();
      await items[itemIndex].click();
      return result.trim();
    }

    throw new Error(
      "Unable to select item '" +
        text +
        "' from " +
        this.selectSelector +
        " list"
    );
  }

  async selectByIndex(index: number) {
    const runInBrowser = function (argument: WebdriverIO.Element) {
      argument.click();
    };
    const select = await this.select;
    await browser.execute(runInBrowser, select);
    await $(OPTION).waitForDisplayed({ timeout: 60000 });
    const item = await $$(OPTION)[index];
    if (item) {
      const result = await item.getText();
      await item.click();
      return result.trim();
    }
    throw new Error(
      "Unable to select index " +
        index +
        " from " +
        this.selectSelector +
        " list"
    );
  }
  async getValue() {
    return await this.select.getValue();
  }
}
