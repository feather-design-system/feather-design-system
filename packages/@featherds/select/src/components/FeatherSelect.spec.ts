import FeatherSelect from "./FeatherSelect.vue";
import { mount } from "@vue/test-utils";
import * as id from "@featherds/utils/id";
import axe from "@featherds/utils/test/axe";
import { getCalls } from "@featherds/utils/test/calls";
import "@featherds/input-helper/test/MutationObserver";
import { nextTick } from "vue";
import { ISelectItemType } from "./types";
import { Code } from "@featherds/utils/keys";
import { vi, expect, describe, it } from "vitest";
vi.spyOn(id, "getSafeId").mockImplementation((x) => x);

const getProps = (props: unknown) => {
  return {
    label: "Users",
    ...(props as Record<string, unknown>),
  };
};

const options = [
  {
    _text: "Item 1",
  },
  {
    _text: "Item 2",
  },
] as ISelectItemType[];
const getFullWrapper = function (options: Record<string, unknown> = {}) {
  document.body.innerHTML = "";
  options.props = getProps(options.props || {});
  options.attachTo = document.body;
  return mount(FeatherSelect, options);
};

describe("FeatherSelect.vue", () => {
  it("should raise the label when it has a value", () => {
    const modelValue = {
      _text: "Test1",
    };
    const wrapper = getFullWrapper({
      props: {
        modelValue,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should not raise the label when its empty", () => {
    const wrapper = getFullWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render in error state", () => {
    const wrapper = getFullWrapper({
      props: {
        error: "ERROR",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render in disabled state", () => {
    const wrapper = getFullWrapper({
      props: {
        disabled: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should open the menu on enter", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.ENTER });

    expect(wrapper.vm.showMenu).toBe(true);
  });
  it("should close the menu on esc and emit update:modelValue", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.ENTER });
    await input.trigger("keydown", { code: Code.ESCAPE });

    expect(wrapper.vm.showMenu).toBe(false);
    expect(
      getCalls<[ISelectItemType]>(wrapper, "update:modelValue")[0][0]
    ).toStrictEqual(options[0]);
  });

  it("should clear the value when clear icon is clicked", async () => {
    const modelValue = {
      _text: "Test1",
    };
    const wrapper = getFullWrapper({
      props: {
        modelValue,
        clear: "Clear Value",
      },
    });
    await wrapper
      .find("[data-ref-id='feather-form-element-clear']")
      .trigger("click");
    expect(
      getCalls<[ISelectItemType]>(wrapper, "update:modelValue")[0][0]
    ).toBeUndefined();
  });
  it("should open and select first item on first down", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.DOWN });

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should open and select first item on first up", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.UP });

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should move selection down when down arrow is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.DOWN });

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[1]);
  });
  it("should not emit update:model till menu is closed", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.DOWN });
    expect(wrapper.vm.showMenu).toBe(true);

    await input.trigger("keydown", { code: Code.ESCAPE });
    expect(wrapper.vm.showMenu).toBe(false);
    expect(
      getCalls<[ISelectItemType]>(wrapper, "update:modelValue")[0][0]
    ).toStrictEqual(options[1]);
  });
  it("should not wrap to the top when down is pressed on the last option", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[options.length - 1],
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.DOWN });

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[options.length - 1]);
  });
  it("should move selection up when up arrow is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[options.length - 1],
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.UP });

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[options.length - 2]);
  });
  it("should not wrap to the bottom when up is pressed on the first option", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.UP });

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should move selection to first item  when home is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[options.length - 1],
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.HOME });
    await nextTick();

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should move selection to last item when end is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.END });

    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[options.length - 1]);
  });
  it("should remain focused when item is selected via click and emit update:modelValue", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.ENTER }); // open menu

    wrapper.findComponent({ ref: "list" }).vm.$emit("select", options[0]);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.showMenu).toBe(false);
    expect(document.activeElement).toBe(wrapper.find("input").element);
    expect(
      getCalls<[ISelectItemType]>(wrapper, "update:modelValue")[0][0]
    ).toStrictEqual(options[0]);
  });
  it("should not open menu when focused (tabbed to)", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    await wrapper.find("input").trigger("focus");
    expect(wrapper.vm.showMenu).toBe(false);
  });
  it("should select item that starts with a character sequence when it is quickly typed", async () => {
    vi.useFakeTimers({
      toFake: ["setTimeout", "clearTimeout"],
    });

    const wrapper = getFullWrapper({
      props: {
        options: options.concat([{ _text: "ab" }]),
      },
    });

    const input = wrapper.find("input");
    await input.trigger("focus");
    await input.trigger("keydown", { code: Code.ENTER });

    await input.trigger("keydown", { key: "a" });
    await input.trigger("keydown", { key: "b" });

    vi.runAllTimers();
    expect(
      (
        wrapper.vm.internalValue as ISelectItemType as { _text: string }
      )._text.indexOf("ab")
    ).toBe(0);
  });
  describe("accessibility", () => {
    it("should be accessible in normal state", async () => {
      vi.useRealTimers();

      const wrapper = getFullWrapper();
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("should be accessible in open state with value", async () => {
      vi.useRealTimers();

      const wrapper = getFullWrapper({
        props: {
          options,
          modelValue: options[0],
        },
      });

      const input = wrapper.find("input");
      await input.trigger("focus");
      await input.trigger("keydown", { code: Code.ENTER });

      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
  });
});
