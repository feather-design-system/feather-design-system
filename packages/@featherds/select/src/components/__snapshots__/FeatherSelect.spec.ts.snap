// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherSelect.vue should not raise the label when its empty 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-ref-id="feather-select-menu-container"
  >
    
    <div
      aria-haspopup="true"
      class="feather-input-wrapper-container feather-select-wrapper"
      id="feather-menu-trigger"
    >
      <div
        class="feather-input-border"
      >
        <div
          class="pre-border"
        />
        <div
          class="label-border"
          style="width: 0px;"
        >
          <label
            class="feather-input-label"
            data-ref-id="feather-form-element-label"
            for="feather-select-input"
          >
            Users
          </label>
        </div>
        <div
          class="post-border"
        />
      </div>
      <div
        class="feather-input-wrapper"
      >
        <div
          class="prefix"
        >
          
          
        </div>
        
        <input
          aria-describedby="feather-select-description"
          aria-disabled="false"
          aria-haspopup="listbox"
          aria-invalid="false"
          class="feather-select-input"
          data-ref-id="feather-select-input"
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
        >
          <!--v-if-->
          <!--v-if-->
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-select-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherSelect.vue should raise the label when it has a value 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-ref-id="feather-select-menu-container"
  >
    
    <div
      aria-haspopup="true"
      class="feather-input-wrapper-container raised feather-select-wrapper"
      id="feather-menu-trigger"
    >
      <div
        class="feather-input-border"
      >
        <div
          class="pre-border"
        />
        <div
          class="label-border"
          style="width: 0px;"
        >
          <label
            class="feather-input-label"
            data-ref-id="feather-form-element-label"
            for="feather-select-input"
          >
            Users
          </label>
        </div>
        <div
          class="post-border"
        />
      </div>
      <div
        class="feather-input-wrapper"
      >
        <div
          class="prefix"
        >
          
          
        </div>
        
        <input
          aria-describedby="feather-select-description"
          aria-disabled="false"
          aria-haspopup="listbox"
          aria-invalid="false"
          class="feather-select-input"
          data-ref-id="feather-select-input"
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
        >
          <!--v-if-->
          <!--v-if-->
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-select-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherSelect.vue should render in disabled state 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-ref-id="feather-select-menu-container"
  >
    
    <div
      aria-haspopup="true"
      class="feather-input-wrapper-container disabled feather-select-wrapper"
      id="feather-menu-trigger"
    >
      <div
        class="feather-input-border"
      >
        <div
          class="pre-border"
        />
        <div
          class="label-border"
          style="width: 0px;"
        >
          <label
            class="feather-input-label"
            data-ref-id="feather-form-element-label"
            for="feather-select-input"
          >
            Users
          </label>
        </div>
        <div
          class="post-border"
        />
      </div>
      <div
        class="feather-input-wrapper"
      >
        <div
          class="prefix"
        >
          
          
        </div>
        
        <input
          aria-describedby="feather-select-description"
          aria-disabled="true"
          aria-haspopup="listbox"
          aria-invalid="false"
          class="feather-select-input"
          data-ref-id="feather-select-input"
          disabled=""
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
        >
          <!--v-if-->
          <!--v-if-->
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-select-description"
  >
    <div
      class="feather-input-spacer"
    >
       
    </div>
    <!--v-if-->
    <!--v-if-->
    
    
  </div>
</div>
`;

exports[`FeatherSelect.vue should render in error state 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-ref-id="feather-select-menu-container"
  >
    
    <div
      aria-haspopup="true"
      class="feather-input-wrapper-container error feather-select-wrapper"
      id="feather-menu-trigger"
    >
      <div
        class="feather-input-border"
      >
        <div
          class="pre-border"
        />
        <div
          class="label-border"
          style="width: 0px;"
        >
          <label
            class="feather-input-label"
            data-ref-id="feather-form-element-label"
            for="feather-select-input"
          >
            Users
          </label>
        </div>
        <div
          class="post-border"
        />
      </div>
      <div
        class="feather-input-wrapper"
      >
        <div
          class="prefix"
        >
          
          
        </div>
        
        <input
          aria-describedby="feather-select-description"
          aria-disabled="false"
          aria-haspopup="listbox"
          aria-invalid="true"
          class="feather-select-input"
          data-ref-id="feather-select-input"
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
        >
          <!--v-if-->
          <svg
            aria-hidden="true"
            class="error-icon hide-when-disabled feather-icon error-icon hide-when-disabled"
            data-ref-id="error-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12,2A10,10,0,1,0,22,12,10,10,0,0,0,12,2Zm0,18a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
            />
            <rect
              height="7"
              rx="1"
              width="2"
              x="11"
              y="7"
            />
            <rect
              height="2"
              rx="0.65"
              width="2"
              x="11"
              y="15"
            />
          </svg>
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="feather-input-sub-text"
    id="feather-select-description"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      aria-live="assertive"
      class="feather-input-error"
      data-ref-id="feather-form-element-error"
    >
      ERROR
    </div>
    
    
  </div>
</div>
`;
